Build on Linux powerpc fails with error: ‘const union ucontext_t::uc_regs_ptr’ has no member named ‘gp_regs’

Building abseil on Linux powerpc (32-bit) fails with:

```
[ 37%] Linking CXX static library libabsl_base.a
cd "/<<PKGBUILDDIR>>/static/absl/base" && /usr/bin/cmake -P CMakeFiles/base.dir/cmake_clean_target.cmake
cd "/<<PKGBUILDDIR>>/static/absl/base" && /usr/bin/cmake -E cmake_link_script CMakeFiles/base.dir/link.txt --verbose=1
/usr/bin/ar qc libabsl_base.a CMakeFiles/base.dir/internal/cycleclock.cc.o CMakeFiles/base.dir/internal/spinlock.cc.o CMakeFiles/base.dir/internal/sysinfo.cc.o CMakeFiles/base.dir/internal/thread_identity.cc.o CMakeFiles/base.dir/internal/unscaledcycleclock.cc.o
In file included from /<<PKGBUILDDIR>>/absl/debugging/stacktrace.cc:46:
/<<PKGBUILDDIR>>/absl/debugging/internal/stacktrace_powerpc-inl.inc: In function ‘void** NextStackFrame(void**, const void*)’:
/<<PKGBUILDDIR>>/absl/debugging/internal/stacktrace_powerpc-inl.inc:134:64: error: ‘const union ucontext_t::uc_regs_ptr’ has no member named ‘gp_regs’; did you mean ‘uc_regs’?
  134 |           reinterpret_cast<void**>(signal_context->uc_mcontext.gp_regs[PT_R1]);
      |                                                                ^~~~~~~
      |                                                                uc_regs
/usr/bin/ranlib libabsl_base.a
make[4]: *** [absl/debugging/CMakeFiles/stacktrace.dir/build.make:85: absl/debugging/CMakeFiles/stacktrace.dir/stacktrace.cc.o] Error 1
make[4]: Leaving directory '/<<PKGBUILDDIR>>/static'
make[3]: *** [CMakeFiles/Makefile2:937: absl/debugging/CMakeFiles/stacktrace.dir/all] Error 2
make[3]: *** Waiting for unfinished jobs....
make[4]: Leaving directory '/<<PKGBUILDDIR>>/static'
```

Looks like the register definitions need to be matched with the glibc headers:

> https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/powerpc/sys/ucontext.h;h=9c57d1103a4312e1aa7e3a8523853348833920fb;hb=HEAD

Full log available in: https://buildd.debian.org/status/fetch.php?pkg=abseil&arch=powerpc&ver=0%7E20200923.3-1&stamp=1612833809&raw=0

Access to a PowerPC big-endian machine running Debian unstable (64 bit) can be obtained through the gcc compile farm: https://gcc.gnu.org/wiki/CompileFarm

--- a/absl/debugging/internal/examine_stack.cc
+++ b/absl/debugging/internal/examine_stack.cc
@@ -56,7 +56,7 @@ void* GetProgramCounter(void* vuc) {
 #elif defined(__powerpc64__)
     return reinterpret_cast<void*>(context->uc_mcontext.gp_regs[32]);
 #elif defined(__powerpc__)
-    return reinterpret_cast<void*>(context->uc_mcontext.regs->nip);
+    return reinterpret_cast<void*>(context->uc_mcontext.uc_regs->gregs[32]);
 #elif defined(__riscv)
     return reinterpret_cast<void*>(context->uc_mcontext.__gregs[REG_PC]);
 #elif defined(__s390__) && !defined(__s390x__)

--- a/absl/debugging/internal/stacktrace_powerpc-inl.inc
+++ b/absl/debugging/internal/stacktrace_powerpc-inl.inc
@@ -131,7 +131,11 @@ static void **NextStackFrame(void **old_sp, const void *uc) {
       const ucontext_t* signal_context =
           reinterpret_cast<const ucontext_t*>(uc);
       void **const sp_before_signal =
+#if defined(__PPC64__)
           reinterpret_cast<void**>(signal_context->uc_mcontext.gp_regs[PT_R1]);
+#else
+          reinterpret_cast<void**>(signal_context->uc_mcontext.uc_regs->gregs[PT_R1]);
+#endif
       // Check that alleged sp before signal is nonnull and is reasonably
       // aligned.
       if (sp_before_signal != nullptr &&

